lib/internal/modules/esm/loader.js

```diff
@@ -168,100 +194,246 @@ class Loader {
     const job = new ModuleJob(this, url, evalInstance, false, false);
     this.moduleMap.set(url, job);
     const { module } = await job.run();
+
     return {
       namespace: module.getNamespace(),
     };
   }
 
-  async import(specifier, parent) {
-    const job = await this.getModuleJob(specifier, parent);
-    const { module } = await job.run();
-    return module.getNamespace();
+  async getModuleJob(specifier, parentURL) {
+    const resolution = await this.resolve(specifier, parentURL);
+    const { url } = resolution;
+    let { format } = resolution;
+    let job = this.moduleMap.get(url);
+    // CommonJS will set functions for lazy job evaluation.
+    if (typeof job === 'function') this.moduleMap.set(url, job = job());
+
+    if (job !== undefined) return job;
+
+    const moduleProvider = async (url, isMain) => {
+      const loaded = await this.load(url, { format });
+      const { source } = loaded;
+      ({ format } = loaded);
```

Comment:
This is actually necessary to avoid an error about using `format` before it's initialised—I tried this first.

Reply:
```suggestion
      const { format: _format, source } = await this.load(url, { format });
      format = _format;
```

Reply:
Cool, I like the less convoluted way here. I think there's an opportunity to be more informative here than `_format` though, perhaps by naming it something like `finalFormat` (since the user can _technically_ return anything from resolve—even an esoteric value like 'typescript').